### DISCUSSION TITLE
Update Digital Ocean marketplace image to latest Mastodon

### DISCUSSION THREAD

**Discussion by cfcode-dev** [mastodon/mastodon#18954] [General]
Would be great to have an up to date Mastodon Digital Ocean marketplace image available to install. 

Current version - 3.1.3 is over 2 years old...I realise doing incremental upgrades all the way to current 3.5.3 is possible, but this looks to be a big job with 4 pages worth of releases and upgrade instructions.

**Comment by cfcode-dev** [comment]
referring to the marketplace app here: https://marketplace.digitalocean.com/apps/mastodon

**Comment by cixastra** [comment]
+1 
Having this issue myself

**Comment by jsailo** [comment]
Have a +1. If the devs aren't bothered about updating the image in DigitalOcean, could someone share tips/links on how to update the version once it's installed by DigitalOcean?
Current image version on DO: 3.1.3 while the latest version os 3.5.3

**Reply by Gargron** [reply]
The instructions are the same for everyone: https://docs.joinmastodon.org/admin/upgrading/

**Reply by cfcode-dev** [reply]
I tried upgrading incrementally from 3.1.3 but ran into error about missing dependencies of some ruby gems, seems the version of Ubuntu (18.04) doesn't have the right software channels setup and after a day of messing around I still couldn't get it working. Deleted the VM and installed from scratch on Ubuntu 20.04 using these instructions. https://www.linode.com/docs/guides/install-mastodon-on-ubuntu-2004/ - working fine now.

I'd consider this DO 'one-click' installation to be pretty useless given it is such an old version and there is no "one-click" upgrade options and you need to be a Mastodon / Ruby expert to upgrade it.

**Reply by jsailo** [reply]
@cfcode-dev - The info on Ubuntu 20.04 is really helpful as the one-click DO installation also installs Mastodon on an Ubuntu 18.04. Also thanks for the Linode link - I had the link bookmarked but having one that is confirmed as working really helps. 

@Gargron - Thanks for the link. But as cfcode-dev has mentioned that the upgrade steps don't work, would be great if someone lets DO know that the one-click image install is deprecated. Just trying to save someone's time.

**Reply by cixastra** [reply]
I tried to contact DigitalOcean support, and they said they would get back to me. I got a reply from a support agent saying that an integrations engineer had been notified of the issue. Then they closed the ticket without leaving me time to reply. >:(

**Comment by cixastra** [comment]
@jsailo @cfcode-dev Hey! I managed to get mastodon installed by using the "Cloudron" image on DigitalOcean's Marketplace. Its sets up and installs mastodon (and many other things) completely automatically. Hope this helps!

 [View Cloudron on the DigitalOcean Marketplace here](https://marketplace.digitalocean.com/apps/cloudron)

**Comment by timnolte** [comment]
Not realizing that the DO version of Mastodon was so out of date, and questioning that it was still running on Ubuntu 18, I finally came across this. I'm attempting to upgrade the Ubuntu distro to a newer version and will then attempt to upgrade Mastodon per the instructions to see how it goes. I'm hoping not to have to start all over again.

**Reply by timnolte** [reply]
> Not realizing that the DO version of Mastodon was so out of date, and questioning that it was still running on Ubuntu 18, I finally came across this. I'm attempting to upgrade the Ubuntu distro to a newer version and will then attempt to upgrade Mastodon per the instructions to see how it goes. I'm hoping not to have to start all over again.

I did successfully use my droplet recovery console to run `/usr/bin/do-release-upgrade` and upgrade to 20.04 LTS. I'm not in the process of the first Mastodon upgrade to 3.2.2. Once I've confirmed the upgrade was successful I'll proceed with additional upgrades. When upgrading to 3.2.2 there are upgrade notes that specifically mention the DO 1-click install and a file that you need to delete in order to run some of the upgrade steps.

**Reply by timnolte** [reply]
> > Not realizing that the DO version of Mastodon was so out of date, and questioning that it was still running on Ubuntu 18, I finally came across this. I'm attempting to upgrade the Ubuntu distro to a newer version and will then attempt to upgrade Mastodon per the instructions to see how it goes. I'm hoping not to have to start all over again.
> 
> I did successfully use my droplet recovery console to run `/usr/bin/do-release-upgrade` and upgrade to 20.04 LTS. I'm not in the process of the first Mastodon upgrade to 3.2.2. Once I've confirmed the upgrade was successful I'll proceed with additional upgrades. When upgrading to 3.2.2 there are upgrade notes that specifically mention the DO 1-click install and a file that you need to delete in order to run some of the upgrade steps.

OK, so for anyone else that does attempt the upgrade process be sure you follow all upgrade steps as defined by each x.x.0 release notes. One issue that I ran into was the need to manually install and update the `Gemfile` to get `mimemagic` installed and then had to run `bundle update mimemagic`. (https://github.com/mastodon/mastodon/issues/16228) In addition I also needed to use Node Source to be able to add Node 14 to the Ubuntu install. `curl -sL https://deb.nodesource.com/setup_14.x -o nodesource_setup.sh`, followed by `bash nodesource_setup.sh`, then ultimately an `apt update` & `apt upgrade` to ensure that Node 14 was installed.

I did have some problems because I missed some of the interim steps, and I also think I was memory restricted on my droplet with only 2GB of RAM. I made a point of shutting down all of Mastodon during the upgrades until it needed it back up. I ended up getting myself to a point where I had to downgrade all the way back to 3.2.2 and the proceed to upgrade all the way back to 3.5.3 before my instance was back running correctly again. As of right now my 1-Click Droplet instance is now at:
* Ubuntu 20.04
* Node 14
* Ruby 3.0.3
* Mastodon 3.5.3
The PostgreSQL instance is still running v10 but v12 is also installed alongside but I have made the migration yet. Once things have settled in I'll take some time to do that.

**Reply by timnolte** [reply]
> > Not realizing that the DO version of Mastodon was so out of date, and questioning that it was still running on Ubuntu 18, I finally came across this. I'm attempting to upgrade the Ubuntu distro to a newer version and will then attempt to upgrade Mastodon per the instructions to see how it goes. I'm hoping not to have to start all over again.
> 
> I did successfully use my droplet recovery console to run `/usr/bin/do-release-upgrade` and upgrade to 20.04 LTS. I'm not in the process of the first Mastodon upgrade to 3.2.2. Once I've confirmed the upgrade was successful I'll proceed with additional upgrades. When upgrading to 3.2.2 there are upgrade notes that specifically mention the DO 1-click install and a file that you need to delete in order to run some of the upgrade steps.

FYI, and additional item that I forgot that came up was the issue mentioned here: https://github.com/mastodon/mastodon/issues/15426#issuecomment-751152190

In the end I think the biggest problem is simply attempting to perform all of the upgrades this late it just showing all of the issues that came up before that have fixes but those aren't listed anywhere so that you can just get through the update. I think we really need to get an updated 1-Click Droplet template that just has 3.5.3 to save everyone all of this trouble.

**Comment by fuzz** [comment]
Is anyone working on this? I am willing to have a go at updating the droplet template, but I don't want to duplicate work.

**Reply by timnolte** [reply]
Looks like this just happened.
 https://mastodon.social/@Gargron/109311569389397185

**Comment by ineffyble** [comment]
DigitalOcean is now updated to 3.5.3, the latest stable release.

**Reply by jleclanche** [reply]
Any idea if it will be updated to the 4.x branch some time soon? And 22.04 LTS instead of 20.04?

**Reply by jsailo** [reply]
@jleclanche I think it's in the hands of DO and not Mastodon. A push from Mastodon's side could surely speed up the process I'm sure.

**Reply by jleclanche** [reply]
Yes it sounded like this from the post... it's weird, because DO claims this is "maintained by Mastodon", not by DO.

**Reply by ineffyble** [reply]
@Gargron how did you work your magic for this before?

**Reply by yatitude** [reply]
I was able to upgrade to my droplet to 4.02 two nights ago with no problems.  Although I may have added one extra release in the process, I was able to use the release notes to upgrade to v3.5.5 then to 4.0.0 then to 4.0.2.  On a very lightly used instance, this took less than 20 minutes. 
Make sure to follow the release notes for each, in particular the one that has you get around upgrading Ruby.

**Reply by timnolte** [reply]
I recently posted my upgrade steps here for reference. https://www.timnolte.com/2022/11/19/upgrading-my-digital-ocean-mastodon-instance-from-3-5-3-to-4-0-2/

**Reply by jsailo** [reply]
@timnolte I'm seeing all JSON data when I visit your link :)

![image](https://user-images.githubusercontent.com/35369823/203604793-9aae4830-e6f5-4275-9819-d7a71772afdb.png)

**Reply by timnolte** [reply]
> @timnolte I'm seeing all JSON data when I visit your link :)
> 
> ![image](https://user-images.githubusercontent.com/35369823/203604793-9aae4830-e6f5-4275-9819-d7a71772afdb.png)

Uh, I think you have something messed up with your browser. I don't have that problem and no one else that has found the post on Mastodon has had any problem. Try copying the URL and use an incognito window without extensions running or a different browser.

**Reply by jsailo** [reply]
@timnolte Weird. I opened the link in Chrome and Chromium too. Same issue. Other links of your site/blog render/open fine. 

https://user-images.githubusercontent.com/35369823/203611757-cf199143-59c8-4118-9146-c3ed90ac0622.mp4

**Reply by ineffyble** [reply]
![Screenshot_20221124_075435_Chrome](https://user-images.githubusercontent.com/834880/203644360-7b927020-95b4-4967-bfa0-253ab017a32e.jpg)

Same here.

**Reply by timnolte** [reply]
OK, it was loading fine on my phone and then checked it again on my desktop and saw the issue. Appears to have been some sort of LiteSpeed cache issue. I purged that page from the cache and all is well now.

**Comment by fsantiago07044** [comment]
so if i'm starting with the DO image with version 3.1.3 of mastodon, what's my best bet to get to 3.5.3 or 4.x? recreate it with your updated marketplace app and import my backed up data created under 3.1.3, or something else?

i already tried the "something else", but may have over complicated and subsequently broken it. i started with do-release-upgrade to get to ubuntu 20.04, and now mastodon lies broken (still at 3.1.3). would appreciate some help from anyone more qualified then I (but i do learn fast if anyone's willing to help me step thru it). i started trying to follow the mastodon upgrade steps, beginning w/ v3.1.4 but error out with (after the git tag checkout steps):

> mastodon@social:~/live$ bundle install
> /home/mastodon/.rbenv/versions/2.6.6/bin/ruby: error while loading shared libraries: libjemalloc.so.1: cannot open shared object file: No such file or directory
> mastodon@social:~/live$
> 

when i tried this before the U 20.04 upgrade (under 18.04), i received a different error relating to the gem file i saw mentioned elsewhere. 

help me masta-DONS, please and thank you.

**Reply by timnolte** [reply]
Hmm, so unfortunately I didn't document my whole process for the OS upgrade and Mastodon upgrade. I did that and managed to get my instance on Ubuntu 20.04 and Mastodon 3.5.3 but it was a painful process. I was going to suggest just migrating to a new updated droplet but there are database migrations that need to happen. You are going to want to proceed with incremental upgrades. 3.1.x to 3.2.x, then 3.2.x to 3.3.x, and so on.

**Comment by fsantiago07044** [comment]
apologies for the strike thru effect over the quoted text.

**Comment by edoswald** [comment]
Just a bit of advice from someone who went through the upgrade process on the DO droplet. What I found is that the upgrade process is very resource intensive. On the smaller droplet sizes, it would continually fail. Going up to the $28/mo level, which I believe is 2 CPUs and 4GB ram, did the trick and was flawless. I'm just going to leave it there as it gives the instance plenty of space to grow, and has been extraordinarily fast in the meantime.

**Reply by fsantiago07044** [reply]
Thanks. My actual os update went fine (just had to fix the apt sources list). But now left with broken mastodon (unsure which piece yet) (and still stuck on rev 3.1.3). Haven't spent much time with it yet to diagnose. Hopefully over the weekend. Thanks.

**Reply by edoswald** [reply]
yeah, try increasing your resources. I saw on Reddit that people said the cheap droplet will work for a small instance, but it's not enough to handle the upgrade processes, so it runs out of memory and fails. So I can't take credit for it. Good luck!

**Reply by fsantiago07044** [reply]
I don't believe it's that. Mine is 2 intel cpu, 4gb ram, 80 Gb ssd. That's gotta be enough. Right now it seems like redis may have failed to start post upgrade but unsure why. There's also an error I mentioned earlier when I attempt a mastodon update. I'll poke at it this weekend.

**Reply by fsantiago07044** [reply]
i'm kind of afraid of how long this will take me to diagnose, especially with someone's prior "painful" comment. i think what i'll do instead is:

1.> re-image my DO droplet to ubuntu 20.04 manually
2.> re-deploy mastodon via the linked-to method using the docker method mentioned earlier on the linode website
3.> just move on from there

luckily, my mastodon instance was very new so i'm not really losing much of anything. so the effort to correct my upgraded droplet seems like maybe not worth it really. plus, who knows what gremlins may remain even if i manage to fix it. 

question for everyone here; would you recommend using docker or other install method?

so, i thank you all and i will report back my progress.

**Comment by fsantiago07044** [comment]
ok done. back up, now on ubuntu 20.04 and mastodon 4.0.2. decided against the docker route and just followed the official mastodon docs. hopefully all done for now.
